pkg/cdi/annotations.go

```diff
@@ -0,0 +1,136 @@
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package cdi
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	// AnnotationPrefix is the prefix for CDI container annotation keys.
+	AnnotationPrefix = "cdi.k8s.io/"
+)
+
+// AnnotateInjection updates annotations with a plugin-specific CDI device
+// injection request for the given devices. Upon any error a non-nil error
+// is returned and annotations are left intact. By convention plugin should
+// be in the format of "vendor.device-type".
+func AnnotateInjection(annotations map[string]string, plugin string, deviceID string, devices []string) (map[string]string, error) {
+	key, err := AnnotationKey(plugin, deviceID)
+	if err != nil {
+		return annotations, errors.Wrap(err, "CDI annotation failed")
+	}
+	if _, ok := annotations[key]; ok {
+		return annotations, errors.Errorf("CDI annotation failed, key %q used", key)
+	}
+	value, err := AnnotationValue(devices)
+	if err != nil {
+		return annotations, errors.Wrap(err, "CDI annotation failed")
+	}
+
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+	annotations[key] = value
+
+	return annotations, nil
+}
+
+// ParseAnnotations parses annotations for CDI device injection requests.
+// The keys and devices from all such requests are collected into slices
+// which are returned as the result. All devices are expected to be fully
+// qualified CDI device names. If any device fails this check empty slices
+// are returned along with a non-nil error. The annotations are expected
+// to be formatted by, or in a compatible fashion to AnnotateInjection().
+func ParseAnnotations(annotations map[string]string) ([]string, []string, error) {
+	var (
+		keys    []string
+		devices []string
+	)
+
+	for key, value := range annotations {
+		if strings.HasPrefix(key, AnnotationPrefix) {
+			for _, d := range strings.Split(value, ",") {
+				if _, _, _, err := ParseQualifiedName(d); err != nil {
+					return nil, nil, err
+				}
+				devices = append(devices, d)
+			}
+			keys = append(keys, key)
+		}
+	}
+
+	return keys, devices, nil
+}
+
+// AnnotationKey returns a unique annotation key for a K8s device plugin.
+// plugin should be in the format of "vendor.device-type". deviceID is the
+// deviceID passed to the device plugin for the allocation.
+func AnnotationKey(plugin, deviceID string) (string, error) {
+	const maxNameLen = 63
+
+	if plugin == "" {
+		return "", errors.New("invalid plugin name, empty")
+	}
+	if deviceID == "" {
+		return "", errors.New("invalid deviceID, empty")
+	}
+
+	name := plugin + "_" + strings.ReplaceAll(deviceID, "/", "_")
```

Comment:
Question: Is there a specific reason that we don't allow `/` in the key? Does replacing them here not make debugging failures more difficult as a user would have to mentally perform this replacement?